File: Algorithms/Medium/DailyTemperatures.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        result = [0] * len(temperatures)
        stack = []

        for i, temp in enumerate(temperatures):
            while stack and temperatures[stack[-1]] < temp:
                prev_idx = stack.pop()
                result[prev_idx] = i - prev_idx
                
            stack.append(i)

        return result

File: Algorithms/Medium/739.Daily-Temperatures/DailyTemperatures.go
package medium

// Time complexity: O(n)
// Space complexity: O(n)
func dailyTemperatures(temperatures []int) []int {
	answer := make([]int, len(temperatures))
	stack := make([]int, 0, len(temperatures)/2)

	pop := func() int {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		return top
	}

	for i, t := range temperatures {
		for len(stack) != 0 && temperatures[stack[len(stack)-1]] < t {
			prev := pop()
			answer[prev] = i - prev
		}

		stack = append(stack, i)
	}

	return answer
}
